gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/api/use_case/UpgradeApiUseCaseTest.java

```diff
@@ -0,0 +1,298 @@
+        userCrudService.initWith(List.of(BaseUserEntity.builder().id(USER_ID).firstname("John").lastname("Doe").build()));
+    }
+
+    @AfterEach
+    void tearDown() {
+        Stream
+            .of(
+                apiCrudService,
+                auditCrudService,
+                indexer,
+                planCrudService,
+                primaryOwnerDomainService,
+                userCrudService,
+                groupQueryService,
+                apiMetadataQueryService,
+                membershipQueryService,
+                roleQueryService,
+                membershipCrudService,
+                apiCategoryQueryService
+            )
+            .forEach(InMemoryAlternative::reset);
+    }
+
+    @Test
+    void should_return_fail_when_api_not_found() {
+        // When
+        var result = useCase.execute(new UpgradeApiUseCase.Input(API_ID, false, EXECUTION_CONTEXT, USER_ID));
+
+        // Then
+        assertThat(result).isInstanceOf(UpgradeApiUseCase.Output.Fail.class);
+        var fail = (UpgradeApiUseCase.Output.Fail) result;
+        assertThat(fail.apiId()).isEqualTo(API_ID);
+        assertThat(fail.errors()).containsExactly("API not found");
+    }
+
+    @Test
+    void should_return_fail_when_api_is_not_v2() {
+        // Given
+        var api = ApiFixtures.aProxyApiV4().toBuilder().id(API_ID).definitionVersion(DefinitionVersion.V4).build();
+        apiCrudService.initWith(List.of(api));
+
+        // When
+        var result = useCase.execute(new UpgradeApiUseCase.Input(API_ID, false, EXECUTION_CONTEXT, USER_ID));
+
+        // Then
+        assertThat(result).isInstanceOf(UpgradeApiUseCase.Output.Fail.class);
+        var fail = (UpgradeApiUseCase.Output.Fail) result;
+        assertThat(fail.apiId()).isEqualTo(API_ID);
+        assertThat(fail.errors()).containsExactly("Cannot upgrade an API which is not a v2 definition");
+    }
+
+    @Test
+    void should_upgrade_api_in_dry_run_mode() {
+        // Given
+        var v2Api = ApiFixtures.aProxyApiV2().toBuilder().id(API_ID).definitionVersion(DefinitionVersion.V2).build();
+        apiCrudService.initWith(List.of(v2Api));
+
+        var plan = PlanFixtures.aPlanV2().toBuilder().id("plan-id").apiId(API_ID).build();
+        planCrudService.initWith(List.of(plan));
+
+        // When
+        var result = useCase.execute(new UpgradeApiUseCase.Input(API_ID, true, EXECUTION_CONTEXT, USER_ID));
+
+        // Then
+        assertThat(result).isInstanceOf(UpgradeApiUseCase.Output.Success.class);
+        var success = (UpgradeApiUseCase.Output.Success) result;
+        assertThat(success.apiId()).isEqualTo(API_ID);
+
+        var storedApi = apiCrudService.findById(API_ID);
+        assertThat(storedApi).map(Api::getDefinitionVersion).hasValue(DefinitionVersion.V2);
+
+        assertThat(auditCrudService.storage()).isEmpty();
+
+        assertThat(indexer.storage()).isEmpty();
+    }
+
+    @Test
+    void should_upgrade_api_and_plans_successfully() {
```

Comment:
```suggestion
    void should_upgrade_api_definition() {
```

And I would add an assertion on the whole API Definition so in this UseCase test we have a complete assertion acting as Acceptance Tests

Reply:
Instead of using Object assertion we could compare JSON 🤔